third_party/move/tools/move-package-manifest/src/manifest.rs

```diff
@@ -0,0 +1,328 @@
+    pub upgrade_policy: Option<UpgradePolicy>,
+}
+
+/// Upgrade policy for a Move package, controlling whether upgrades are allowed and if so, what rules to follow.
+#[derive(Debug, Clone, Eq, PartialEq, Serialize, Deserialize)]
+pub enum UpgradePolicy {
+    /// Indicates that upgrades are allowed, but must be compatible with the current version.
+    #[serde(rename = "compatible")]
+    Compatible,
+
+    /// Indicates that the package is immutable and should not be upgraded.
+    #[serde(rename = "immutable")]
+    Immutable,
+}
+
+/// A semantic version consisting of major, minor, and patch components.
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct Version {
+    pub major: u64,
+    pub minor: u64,
+    pub patch: u64,
+}
+
+/// Represents either a wildcard or a numerical address.
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub enum AddressOrWildcard {
+    /// Wildcard address (`"_"` in `Move.toml`).
+    Wildcard,
+
+    /// A specific numerical address.
+    Numerical(AccountAddress),
+}
+
+/// Build options defined in the `[build]` section of `Move.toml`.
+#[derive(Default, Debug, Clone, Eq, PartialEq, Serialize, Deserialize)]
+#[serde(deny_unknown_fields)]
+pub struct BuildOptions {
+    pub language_version: Option<Version>,
+}
+
+/// Represents a dependency entry in `[dependencies]` or `[dev-dependencies]`.
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct Dependency {
+    /// Optional version requirement for the dependency.
+    version: Option<Version>,
+
+    /// Location of the dependency: local, git, or aptos (on-chain).
+    location: PackageLocation,
+}
+
+/// Location of a package dependency.
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub enum PackageLocation {
+    /// Refers to a package stored in the local file system.
+    Local { path: String },
+
+    /// Refers to a package stored in a git repository.
+    Git {
+        /// URL to the Git repository.
+        url: Url,
+        /// Optional Git revision to pin the dependency to.
+        rev: Option<String>,
+        /// Optional subdirectory within the Git repository.
+        subdir: Option<String>,
+    },
+
+    /// Refers to a package published on-chain.
+    // REVISIT: Potentially leaky abstraction -- do we still want this to be platform agnostic?
+    Aptos {
+        /// URL to the Aptos full-node connected to the network where the package is published.
+        node_url: String,
+
+        /// Address of the published package.
+        package_addr: String,
+    },
+}
+
+/***************************************************************************************************
+ * Custom Deserializer Implementations
+ *
+ **************************************************************************************************/
+impl<'de> Deserialize<'de> for Version {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        let make_err = || serde::de::Error::custom("Invalid version format");
```

Comment:
the error message in `parse_version` of `move-package/src/source_package/manifest_parser.rs` seems much more robust error message for malformed version format. Do we want to keep this robust error message?

Reply:
Good catch. I got the message improved.